File: poker/card.py
from __future__ import annotations

import random
from enum import IntEnum


class Suit(IntEnum):
    CLUBS = 0
    DIAMONDS = 1
    HEARTS = 2
    SPADES = 3


SUIT_SYMBOLS = {
    Suit.CLUBS: "♣",
    Suit.DIAMONDS: "♦",
    Suit.HEARTS: "♥",
    Suit.SPADES: "♠",
}

SUIT_COLORS = {
    Suit.CLUBS: (30, 30, 30),
    Suit.DIAMONDS: (200, 30, 30),
    Suit.HEARTS: (200, 30, 30),
    Suit.SPADES: (30, 30, 30),
}

RANK_NAMES = {
    2: "2", 3: "3", 4: "4", 5: "5", 6: "6", 7: "7", 8: "8",
    9: "9", 10: "10", 11: "J", 12: "Q", 13: "K", 14: "A",
}


class Card:
    __slots__ = ("rank", "suit")

    def __init__(self, rank: int, suit: Suit):
        self.rank = rank
        self.suit = suit

    @property
    def symbol(self) -> str:
        return SUIT_SYMBOLS[self.suit]

    @property
    def rank_str(self) -> str:
        return RANK_NAMES[self.rank]

    @property
    def color(self) -> tuple:
        return SUIT_COLORS[self.suit]

    @property
    def is_red(self) -> bool:
        return self.suit in (Suit.HEARTS, Suit.DIAMONDS)

    def __repr__(self) -> str:
        return f"{self.rank_str}{self.symbol}"

    def __eq__(self, other) -> bool:
        if not isinstance(other, Card):
            return NotImplemented
        return self.rank == other.rank and self.suit == other.suit

    def __hash__(self) -> int:
        return hash((self.rank, self.suit))

    def __lt__(self, other) -> bool:
        if not isinstance(other, Card):
            return NotImplemented
        return self.rank < other.rank


class Deck:
    def __init__(self):
        self.cards: list[Card] = []
        self.reset()

    def reset(self):
        self.cards = [
            Card(rank, suit)
            for suit in Suit
            for rank in range(2, 15)
        ]
        self.shuffle()

    def shuffle(self):
        random.shuffle(self.cards)

    def deal(self, count: int = 1) -> list[Card]:
        dealt = self.cards[:count]
        self.cards = self.cards[count:]
        return dealt

    def deal_one(self) -> Card:
        return self.cards.pop(0)

    def __len__(self) -> int:
        return len(self.cards)

File: TexasPoker/Models/Card.swift
import Foundation

enum Suit: Int, CaseIterable, Comparable, Codable {
    case clubs = 0
    case diamonds = 1
    case hearts = 2
    case spades = 3

    var symbol: String {
        switch self {
        case .clubs: return "♣"
        case .diamonds: return "♦"
        case .hearts: return "♥"
        case .spades: return "♠"
        }
    }

    var isRed: Bool {
        self == .hearts || self == .diamonds
    }

    static func < (lhs: Suit, rhs: Suit) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

struct Card: Identifiable, Equatable, Comparable, Hashable {
    let id = UUID()
    let rank: Int   // 2-14, where 14 = Ace
    let suit: Suit

    var rankString: String {
        switch rank {
        case 2...10: return "\(rank)"
        case 11: return "J"
        case 12: return "Q"
        case 13: return "K"
        case 14: return "A"
        default: return "?"
        }
    }

    var displayName: String {
        "\(rankString)\(suit.symbol)"
    }

    static func < (lhs: Card, rhs: Card) -> Bool {
        if lhs.rank != rhs.rank { return lhs.rank < rhs.rank }
        return lhs.suit < rhs.suit
    }

    static func == (lhs: Card, rhs: Card) -> Bool {
        lhs.rank == rhs.rank && lhs.suit == rhs.suit
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(rank)
        hasher.combine(suit)
    }
}

class Deck {
    private var cards: [Card] = []

    init() {
        reset()
    }

    func reset() {
        cards = []
        for suit in Suit.allCases {
            for rank in 2...14 {
                cards.append(Card(rank: rank, suit: suit))
            }
        }
        shuffle()
    }

    func shuffle() {
        cards.shuffle()
    }

    func deal(_ count: Int = 1) -> [Card] {
        let dealt = Array(cards.prefix(count))
        cards.removeFirst(min(count, cards.count))
        return dealt
    }

    func dealOne() -> Card? {
        cards.isEmpty ? nil : cards.removeFirst()
    }

    var remaining: Int { cards.count }
}
